### DISCUSSION TITLE
Using PnP.PowerShell in Azure runbook fails

### DISCUSSION THREAD

**Discussion by colonelclaypoo** [pnp/powershell#467] [General] [answered]
Hello,

I'm trying to execute the latest PnP.PowerShell in an Azure runbook via an Azure app.
The authentication seems to work just fine. However, the cmdlet fails with the error "User not found" - for whatever reason.

```
$connection = Get-AutomationConnection –Name "AzureRunAsConnection"
$tenant = '[tenantID]'
$url = 'https://[tenant].sharepoint.com/sites/test'

Connect-PnPOnline -Url $url –Thumbprint "[thumbprint]" –ClientId "[clientID]" -Tenant $tenant -ReturnConnection

Get-PnPWeb
```

```
Get-PnPWeb : User cannot be found.
At line:8 char:1
+ Get-PnPWeb
+ ~~~~~~~~~~
    + CategoryInfo          : WriteError: (:) [Get-PnPWeb], ServerException
    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.GetWeb
```

I tried storing the cmdlet output in a variable and what happens then is that the runbook times out and gets suspended after 3 unsuccessful attempts. The time out happens for other cmdlets, such as Get-PnPTenantSite even without storing them in a variable. Piping into | Out-Null also times out.

The same method (auth via Azure app with certificate) works on a local machine, though.
ExchangeV2 cmdlets work in this environment with the described method.

Anyone observed this too?
Thanks.

**Comment by AndersRask** [comment, marked as answer]
Always store all cmdlet calls in variables or use out-null to prevent runbooks from getting suspended and always create runbooks logic as idempotent (it is the build in retry logic that stems from when they were used as workflows that kicks in) 

Did you assign O365 graph permissions to the runas service principal in AAD?

Kind regards Anders Rask 

On 20 Mar 2021 08.15, colonelclaypoo ***@***.***> wrote:

Hello,

I'm trying to execute the latest PnP.PowerShell in an Azure runbook via an Azure app.

The authentication seems to work just fine. However, the cmdlet fails with the error "User not found" - for whatever reason.

$connection = Get-AutomationConnection –Name "AzureRunAsConnection"

$tenant = '[tenantID]'

$url = 'https://[tenant].sharepoint.com/sites/test'



Connect-PnPOnline -Url $url –Thumbprint "[thumbprint]" –ClientId "[clientID]" -Tenant $tenant -ReturnConnection



Get-PnPWeb



Get-PnPWeb : User cannot be found.

At line:8 char:1

+ Get-PnPWeb

+ ~~~~~~~~~~

    + CategoryInfo          : WriteError: (:) [Get-PnPWeb], ServerException

    + FullyQualifiedErrorId : EXCEPTION,PnP.PowerShell.Commands.GetWeb



I tried storing the cmdlet output in a variable and what happens then is that the runbook times out and gets suspended after 3 unsuccessful attempts. The time out happens for other cmdlets, such as Get-PnPTenantSite even without storing them in a variable. Piping into | Out-Null also times out.

The same method (auth via Azure app with certificate) works on a local machine, though.

ExchangeV2 cmdlets work in this environment with the described method.

Anyone observed this too?

Thanks.
